### PR TITLE
Enforce consent

### PR DESCRIPTION
**Pull request by addb**
### Edge Request Worker: Queue Processing
Screenshot 2024-08-23 at 12 22 56 PM

### Consent: setConsent API flow
Screenshot 2024-08-26 at 5 52 00 PM

### Consent Module working
Screenshot 2024-08-26 at 5 55 42 PM

### Changes
- Enforce consent in EdgeModule using consentState
- Refactor EdgeRequestWorker to use EdgeRequest 
- Add EdgeRequest class
- Add tests for Edge request and consent enforcement in EdgeRequestWorker
- Update tests
- Enforce consent in IdentityModule using consentState
- Update and add tests for Identity

### Extra changes
- Use helpers in Media classes 
- logging improvements and cleanup
- fix unused variables warning by prefixing `'_'`

 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: code/main/edge/edgeModule.brs [hunk 1]
```diff
@@ -28,22 +28,33 @@ function _adb_EdgeModule(configurationModule as object, identityModule as object
...
 
         ' sendEvent API triggers this API to queue edge requests
         ' requestId: unique id for the request
         ' eventData: data to be sent to edge
         ' timestampInMillis: timestamp of the event
-        processEvent: function(requestId as string, eventData as object, timestampInMillis as longinteger) as void
-            m._edgeRequestWorker.queue(requestId, eventData, timestampInMillis, {}, m._EDGE_REQUEST_PATH)
+        processEvent: function(requestId as string, eventData as object, timestampInMillis as longinteger, requestType = m._REQUEST_TYPE_EDGE as string) as void
+            edgeRequest = _adb_EdgeRequest(requestId, eventData, timestampInMillis)
+            edgeRequest.setMeta({})
```
Comment (praveek) [inline code/main/edge/edgeModule.brs:53]: nit: Is this necessary? Can't you handle it when processing the request?

Comment (addb) [reply code/main/edge/edgeModule.brs:53]: yes, we do update it while processing

File: code/main/edge/edgeRequestWorker.brs [hunk 2]
```diff
@@ -251,8 +383,43 @@ function _adb_EdgeRequestWorker(edgeResponseManager as object) as object
...
+
+            if _adb_stringEqualsIgnoreCase(collectConsent, m._COLLECT_CONSENT_YES)
+                _adb_logVerbose("EdgeRequestWorker::_isBlockedByConsent() - Collect consent value is set to (" + FormatJson(collectConsent) + "). The edge requests will not be blocked.")
+                return false
+            end if
+
+            if _adb_stringEqualsIgnoreCase(collectConsent, m._COLLECT_CONSENT_NO)
+                _adb_logVerbose("EdgeRequestWorker::_isBlockedByConsent() - Collect consent value is set to (" + FormatJson(collectConsent) + "). The edge requests will not be blocked.")
+                return false
+            end if
```
Comment (kevinlind) [inline code/main/edge/edgeRequestWorker.brs:414]: You can join these into a single `if`.

File: code/main/edge/edgeRequestWorker.brs [hunk 3]
```diff
@@ -210,19 +312,49 @@ function _adb_EdgeRequestWorker(edgeResponseManager as object) as object
...
+
+        _createConsentRequestBody: function(consentData as object, ecid as string, meta as object) as object
+            requestBody = {
+                "query" : {
+                    "consent" : {
+                    "operation" : "update"
+                    }
+                },
+                "xdm": {
+                    "identityMap": m._getIdentityMap(ecid),
```
Comment (kevinlind) [inline code/main/edge/edgeRequestWorker.brs:344]: Do you need the same logic here as done in `_createEdgeRequestBody` to check if `ecid` is invalid or empty? Will Konductor generate and return an ECID value for update consent requests?

Comment (addb) [reply code/main/edge/edgeRequestWorker.brs:344]: Yes, we will be removing ECID requirement later, I can revert this change for now, but our logic will not work since identity is blocked by consent and we might not have ECID. This can be a part of new task #188

**Review by kevinlind — APPROVED** [review]

File: code/main/media/mediaModule.brs [hunk 4]
```diff
@@ -51,7 +51,7 @@ function _adb_MediaModule(configurationModule as object, edgeModule as object) a
             eventType = eventData.xdmData.xdm.eventType
             clientSessionId = eventData.clientSessionId
 
-            if eventType <> m._CONSTANTS.MEDIA.EVENT_TYPE.SESSION_START and clientSessionId <> m._sessionManager.getActiveClientSessionId() then
+            if not _adb_stringEqualsIgnoreCase(eventType, m._CONSTANTS.MEDIA.EVENT_TYPE.SESSION_START) and clientSessionId <> m._sessionManager.getActiveClientSessionId() then
```
Comment (praveek) [inline code/main/media/mediaModule.brs:54]: There should be an `or` separating these two checks.

File: code/main/edge/edgeRequestWorker.brs [hunk 5]
```diff
@@ -87,72 +89,172 @@ function _adb_EdgeRequestWorker(edgeResponseManager as object) as object
...
                 end if
 
                 ' grab oldest hit in the queue
-                requestEntity = m._queue.Shift()
+                requestToBeSent = invalid
+
+                if m._isBlockedByConsent(m._consentState)
+                    ' since the consent is pending process the consent request
+                    consentRequest = m._consentQueue.Shift()
+                    if not _adb_isEdgeConsentRequest(consentRequest)
```
Comment (praveek) [inline code/main/edge/edgeRequestWorker.brs:107]: This section will be more readable if you just check 
```
if consentQueue is empty 
	exit processing
```

File: code/main/media/mediaModule.brs (hunk 4 above)
Comment (addb) [reply code/main/media/mediaModule.brs:56]: It is supposed to be and. Added comments to clarify the logic.

File: code/main/edge/edgeRequestWorker.brs (hunk 5 above)
Comment (addb) [reply code/main/edge/edgeRequestWorker.brs:107]: okay

File: code/main/edge/edgeRequestWorker.brs [hunk 6]
```diff
@@ -87,72 +89,172 @@ function _adb_EdgeRequestWorker(edgeResponseManager as object) as object
...
+                        exit while
+                    end if
+
+                    requestToBeSent = consentRequest
+                else
+                    ' process the oldest request from both queues
+                    edgeRequest = m._queue.Shift() ' get the oldest edge request
+                    consentRequest = m._consentQueue.Shift() ' get the oldest consent request
+
+                    if not _adb_isEdgeConsentRequest(consentRequest) ' consent request might be invalid if queue is empty
```
Comment (praveek) [inline code/main/edge/edgeRequestWorker.brs:120]: The code will be readable if you just check if the queue is empty rather than getting the oldest request and checking if it is invalid.

File: code/main/edge/edgeRequestWorker.brs [hunk 7]
```diff
@@ -87,72 +89,172 @@ function _adb_EdgeRequestWorker(edgeResponseManager as object) as object
...
+                        requestToBeSent = edgeRequest
+                    else if not _adb_isValidEdgeRequest(edgeRequest) ' edge request might be invalid if queue is empty
+                        requestToBeSent = consentRequest
```
Comment (praveek) [inline code/main/edge/edgeRequestWorker.brs:123]: I’m not clear on where you exit the loop after sending the consentRequest, ensuring that the consentState is updated.

File: code/main/edge/edgeRequestWorker.brs (hunk 6 above)
Comment (addb) [reply code/main/edge/edgeRequestWorker.brs]: @praveek updated the logic let me know if this seems better.

File: code/main/edge/edgeRequestWorker.brs [hunk 8]
```diff
@@ -87,72 +89,175 @@ function _adb_EdgeRequestWorker(edgeResponseManager as object) as object
...
+                collectConsent = m._consentState.getCollectConsent()
+                ' If collect consent is set to "n" and the request is not a consent request, drop the request
+                if not _adb_isEdgeConsentRequest(requestToBeSent) and _adb_stringEqualsIgnoreCase(collectConsent, m._COLLECT_CONSENT_NO)
+                    _adb_logVerbose("EdgeRequestWorker::processRequests() - Collect consent value is set to (" + FormatJson(collectConsent) + "). The Edge request with id:(" + FormatJson(requestToBeSent.getRequestId()) + ") will be dropped.")
                     continue while
                 end if
 
-                _adb_logVerbose("EdgeRequestWorker::processRequests() - Request with id:(" + FormatJson(requestId) + ") response: " + networkResponse.toString())
+                networkResponse = m._processRequest(edgeConfig, requestToBeSent)
+                edgeResponse = m._processResponse(networkResponse, requestToBeSent)
```
Comment (praveek) [inline code/main/edge/edgeRequestWorker.brs:151]: I don't seem to follow where the the consentState is updated with consent response?

Comment (addb) [reply code/main/edge/edgeRequestWorker.brs:151]: @praveek 
consentState is updated only by consentModule and only based on response.

eventProcessor triggers this function i.e. EdgeRequestWorker.processRequests() and get's array of responses which it forwards to modules like consent, media, etc.  Consent handles the consent:preferences handle and updates the state.

Comment (addb) [reply code/main/edge/edgeRequestWorker.brs:151]: @praveek updated description with the diagram.

Comment (addb) [reply code/main/edge/edgeRequestWorker.brs:151]: @praveek Updated to logic to break the loop after sending out consent request to facilitate consent processing.

**Review by praveek — APPROVED** [review]
